### PR TITLE
game_phase_tasks

### PR DESCRIPTION
**Pull request by TheEmidee**
- Made functions that observe phase start or end return a handle
- Added BP tasks to observe the start or end of game phases

- - -
This change is [Reviewable](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219)

### CODE REVIEW

File: Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp [hunk 1]
```diff
@@ -247,14 +264,23 @@ void UGBFGamePhaseSubsystem::OnBeginPhase( const UGBFGamePhaseAbility * phase_ab
         auto & entry = ActivePhaseMap.FindOrAdd( phase_ability_handle );
         entry.PhaseTag = incoming_phase_tag;
 
+        TArray< FGBFGamePhaseObserverHandle > observer_indices_to_remove;
+
         // Notify all observers of this phase that it has started.
-        for ( const auto & observer : PhaseStartObservers )
+        for ( auto index = 0; index < PhaseStartObservers.Num(); ++index )
```
Comment (Frickerson) [inline Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp]: could keep the foreach?

File: Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp [hunk 2]
```diff
@@ -268,14 +294,41 @@ void UGBFGamePhaseSubsystem::OnEndPhase( const UGBFGamePhaseAbility * phase_abil
 
     ActivePhaseMap.Remove( phase_ability_handle );
 
+    TArray< FGBFGamePhaseObserverHandle > observer_indices_to_remove;
+
     // Notify all observers of this phase that it has ended.
-    for ( const auto & observer : PhaseEndObservers )
+    for ( auto index = 0; index < PhaseEndObservers.Num(); ++index )
```
Comment (Frickerson) [inline Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp]: Same

**Review by Frickerson — CHANGES_REQUESTED** [review]

**Review by TheEmidee — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @VectorGamez)

___
*[`Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp` line 270 at r1](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219#gh-1447109711:-NnonvXiAqja6LXHnTfc:bvd93x1) ([raw file](https://github.com/TheEmidee/UEGameBaseFramework/blob/a6c74bdc24e6088b94c886d93d525627477f0728/Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp#L270)):*
<details><summary><i>Previously, VectorGamez (Viktor Blomme) wrote…</i></summary><blockquote>

could keep the foreach?
</blockquote></details>

I need the index

**Review by TheEmidee — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219)* status: 0 of 4 files reviewed, 2 unresolved discussions (waiting on @TheEmidee and @VectorGamez)

___
*[`Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp` line 300 at r1](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219#gh-1447110169:-NnonyIXE8C9zLyWeKVQ:b-896fix) ([raw file](https://github.com/TheEmidee/UEGameBaseFramework/blob/a6c74bdc24e6088b94c886d93d525627477f0728/Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp#L300)):*
<details><summary><i>Previously, VectorGamez (Viktor Blomme) wrote…</i></summary><blockquote>

Same
</blockquote></details>

Done.

**Review by Frickerson — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219)* status: all files reviewed, 2 unresolved discussions (waiting on @TheEmidee)

___
*[`Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp` line 270 at r1](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219#gh-1447109711:-NnrZUrm7BtBWwwFjsyw:b5al5sd) ([raw file](https://github.com/TheEmidee/UEGameBaseFramework/blob/a6c74bdc24e6088b94c886d93d525627477f0728/Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp#L270)):*
<details><summary><i>Previously, TheEmidee (Michael Delva) wrote…</i></summary><blockquote>

I need the index
</blockquote></details>

It does not seem to be used right now though?

**Review by TheEmidee — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219)* status: all files reviewed, 2 unresolved discussions (waiting on @VectorGamez)

___
*[`Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp` line 270 at r1](https://reviewable.io/reviews/TheEmidee/UEGameBaseFramework/219#gh-1447109711:-NnrrLpd6-9hmg6fEbKZ:b-896fix) ([raw file](https://github.com/TheEmidee/UEGameBaseFramework/blob/a6c74bdc24e6088b94c886d93d525627477f0728/Source/GameBaseFramework/Private/Phases/GBFGamePhaseSubsystem.cpp#L270)):*
<details><summary><i>Previously, VectorGamez (Viktor Blomme) wrote…</i></summary><blockquote>

It does not seem to be used right now though?
</blockquote></details>

Done.

**Pull request merged**